### PR TITLE
[XLA:GPU][ROCm] AMDGPU-specific logic to produce HSA code objects for XLA.

### PR DESCRIPTION
**Pull request by whchung**
Follow-up to #30326  .

- rename `nvptx_backend_lib` to `gpu_backend_lib`.
- introduce `xla::gpu::amdgpu::CompileToHsaco`.
- AMDGPU-specific logic to drive LLVM / LLD to produce HSA code objects which are used on AMD ROCm platform.

### CODE REVIEW

File: tensorflow/compiler/xla/service/gpu/llvm_gpu_backend/gpu_backend_lib.cc [hunk 1]
```diff
@@ -558,6 +560,94 @@ static std::vector<string> GetROCDLPaths(int amdgpu_version,
...
+// TargetMachine for the AMDGPU target.
+StatusOr<std::vector<uint8>> EmitModuleToHsaco(
+    Module* module, llvm::TargetMachine* target_machine) {
+  char tempdir_template[] = "/tmp/amdgpu_xla-XXXXXX";
+  char* tempdir_name = mkdtemp(tempdir_template);
+
+  VLOG(1) << "Compile-time artifacts located at: " << tempdir_name;
+
+  // prepare filenames for all stages of compilation:
+  // IR, ISA, binary ISA, and HSACO
```
Comment (thomasjoerg) [inline tensorflow/compiler/xla/service/gpu/llvm_gpu_backend/gpu_backend_lib.cc]: The code below sets up filenames for isabin (binary ISA) but not for ISA alone.

File: tensorflow/compiler/xla/service/gpu/llvm_gpu_backend/gpu_backend_lib.cc [hunk 2]
```diff
@@ -558,6 +560,94 @@ static std::vector<string> GetROCDLPaths(int amdgpu_version,
...
+               << lld_program.getError().message();
+  }
+  std::vector<llvm::StringRef> lld_args{
+      llvm_ir::AsStringRef("ld.lld"),      llvm_ir::AsStringRef("-flavor"),
+      llvm_ir::AsStringRef("gnu"),         llvm_ir::AsStringRef("-shared"),
+      llvm_ir::AsStringRef("isabin_path"), llvm_ir::AsStringRef("-o"),
+      llvm_ir::AsStringRef("hsaco_path"),
+  };
+  lld_args[4] = llvm_ir::AsStringRef(isabin_path.c_str());
+  lld_args[6] = llvm_ir::AsStringRef(hsaco_path.c_str());
```
Comment (thomasjoerg) [inline tensorflow/compiler/xla/service/gpu/llvm_gpu_backend/gpu_backend_lib.cc]: Please initialize lld_args[4] and [6] in place.

File: tensorflow/compiler/xla/service/gpu/llvm_gpu_backend/gpu_backend_lib.cc [hunk 3]
```diff
@@ -591,7 +681,63 @@ std::unique_ptr<llvm::TargetMachine> AMDGPUGetTargetMachine(
...
+        AMDGPUGetTargetMachine(default_target_triple, *amdgpu_version,
+                               hlo_module_config);
+
+    // Link with ROCm-Device-Libs, and optimize the LLVM module.
+    TF_RETURN_IF_ERROR(LinkAndOptimizeModule(
+        module, gpu_version, hlo_module_config, rocdl_dir_path,
+        AMDGPUTargetModuleLinker, default_target_triple, target_machine.get(),
+        kAMDGPUInlineThreshold));
+
+    // Lower optimize LLVM module to HSA code object.
```
Comment (thomasjoerg) [inline tensorflow/compiler/xla/service/gpu/llvm_gpu_backend/gpu_backend_lib.cc]: Typo: optimized

**Review by thomasjoerg — DISMISSED** [review]

File: tensorflow/compiler/xla/service/gpu/llvm_gpu_backend/gpu_backend_lib.cc [hunk 4]
```diff
@@ -558,6 +560,94 @@ static std::vector<string> GetROCDLPaths(int amdgpu_version,
   return result;
 }
 
+// Emits the given module to HSA Code Object. target_machine is an initialized
+// TargetMachine for the AMDGPU target.
+StatusOr<std::vector<uint8>> EmitModuleToHsaco(
+    Module* module, llvm::TargetMachine* target_machine) {
+  char tempdir_template[] = "/tmp/amdgpu_xla-XXXXXX";
```
[the remark this answers is not in the document]
Comment (whchung) [reply tensorflow/compiler/xla/service/gpu/llvm_gpu_backend/gpu_backend_lib.cc]: thanks. will revise to use specified API.

File: tensorflow/compiler/xla/service/gpu/llvm_gpu_backend/gpu_backend_lib.cc (hunk 1 above)
Comment (whchung) [reply tensorflow/compiler/xla/service/gpu/llvm_gpu_backend/gpu_backend_lib.cc]: will revise the comment.

File: tensorflow/compiler/xla/service/gpu/llvm_gpu_backend/gpu_backend_lib.cc [hunk 5]
```diff
@@ -558,6 +560,94 @@ static std::vector<string> GetROCDLPaths(int amdgpu_version,
...
+      tensorflow::io::JoinPath(tempdir_name, isabin_filename);
+
+  std::string hsaco_filename =
+      absl::StrCat(module->getModuleIdentifier(), ".hsaco");
+  std::string hsaco_path =
+      tensorflow::io::JoinPath(tempdir_name, hsaco_filename);
+
+  std::error_code ec;
+
+  // dump LLVM IR
```
[the remark this answers is not in the document]
Comment (whchung) [reply tensorflow/compiler/xla/service/gpu/llvm_gpu_backend/gpu_backend_lib.cc]: will revise.

File: tensorflow/compiler/xla/service/gpu/llvm_gpu_backend/gpu_backend_lib.cc (hunk 2 above)
Comment (whchung) [reply tensorflow/compiler/xla/service/gpu/llvm_gpu_backend/gpu_backend_lib.cc]: will revise.

File: tensorflow/compiler/xla/service/gpu/llvm_gpu_backend/gpu_backend_lib.cc (hunk 3 above)
Comment (whchung) [reply tensorflow/compiler/xla/service/gpu/llvm_gpu_backend/gpu_backend_lib.cc]: will revise.

**Review by thomasjoerg — APPROVED** [review]

**Comment by thomasjoerg** [comment]
@whchung 
We need to resolve some internal build setup issue before merging this PR. Sorry to make you wait. I hope this won't take long.

**Comment by whchung** [comment]
@thomasjoerg @chsigg any chance I can understand how this PR broke XLA for CUDA?

**Comment by whchung** [comment]
@thomasjoerg / @chsigg ping. may I understand how this PR broke XLA for CUDA?

**Comment by chsigg** [comment]
An internal test assumed the file you renamed in this PR to contain the string 'ptx'. I've fixed it and the PR should be merged again shortly.

**Pull request merged**